### PR TITLE
Add pod annotations

### PR DESCRIPTION
**Pull request by Kasama**
Add support for annotations by using the config variable `KUBERNETES_PODS_ANNOTATIONS`. This variable should be a json containing the `app_type` as the key and the annotations as an object.

Example with `app_type = cmd` 

```json
{
  "cmd": {
    "k8s.annotation/sample": "value",
    "k8s.another/annotation":"anotherValue"
  }
}
```

### CODE REVIEW

**Comment by felipejfc** [comment]
Hi @Kasama, I think it would be better this way:

```
dprod annotation -h

annotation:list        list annotations for an app
annotation:set         set annotations for an app
annotation:unset       unset annotations for an app
```
this would be very like the config command:

```
dprod config -h

Valid commands for config:

config:list        list environment variables for an app
config:set         set environment variables for an app
config:unset       unset environment variables for an app
...
```

So if I want to set annotation k8s/annotation=value to container metagame of app "leapp" I'd use:

```
dprod annotation:set k8s/annotation=value -a leapp -c metagame
```

what do you think?

**Comment by Kasama** [comment]
@felipejfc, although I do agree with you that it would be more user-friendly to have a command for that in the `cli`, me and @cscatolini spoken about this before I begun and we had decided that for now it would be best to make this simple change instead of adding a new command to the `cli`

**Comment by cscatolini** [comment]
@felipejfc we've decided to follow the same approach used to support tolerations (via env vars) so there would be changes only on the controller instead of changing the controller, the cli and the builder. This will allow us to fix this issue quicker, since it is already impacting our deis deployments running with the new datadog agent.

File: rootfs/api/models/app.py [hunk 1]
```diff
@@ -1111,6 +1111,14 @@ def _gather_app_settings(self, release, app_settings, process_type, replicas):
                 self.log(err, logging.ERROR)
                 tolerations = {}
 
+        # set pod annotations
+        annotations = config.values.get('KUBERNETES_PODS_ANNOTATIONS', settings.KUBERNETES_PODS_ANNOTATIONS) # noqa
+        if annotations:
+            try:
+                annotations = json.loads(annotations)
+            except:
```
Comment (cscatolini) [inline rootfs/api/models/app.py]: I'd rather have an explicit exception here

File: rootfs/scheduler/resources/pod.py [hunk 2]
```diff
@@ -175,20 +180,20 @@ def manifest(self, namespace, name, image, **kwargs):
...
         try:
-            app_type = kwargs.get('app_type')
-            tolerations = kwargs.get('tolerations', {})
-            return tolerations.get(app_type)
+            app_name = kwargs.get('app_type')
+            json = kwargs.get(json_name, {})
+            return json.get(app_name)
         except:
             return None
+        return ""
```
Comment (cscatolini) [inline rootfs/scheduler/resources/pod.py]: this return is unnecessary, we'll either fall in the try or the except returns

File: rootfs/scheduler/resources/pod.py [hunk 3]
```diff
@@ -120,6 +120,11 @@ def manifest(self, namespace, name, image, **kwargs):
             'spec': {}
         }
 
+        # pod annotations
+        annotations = self._get_json_from_app('pod_annotations', **kwargs)
+        if annotations:
+            manifest['metadata']['annotations'] = annotations
```
Comment (cscatolini) [inline rootfs/scheduler/resources/pod.py:126]: we need to ensure we're not over-writting any of deis annotations, are there any tests for it?

**Review by cscatolini — CHANGES_REQUESTED** [review]
Some small changes and a check before proceeding.

File: rootfs/scheduler/resources/pod.py (hunk 3 above)
Comment (Kasama) [reply rootfs/scheduler/resources/pod.py:126]: Deis itself does not add any annotations to the `pod` template metadata, it adds the `kubernetes.io/change-cause` and `deployment.kubernetes.io/revision` annotations to the `deployment` metadata only, so i don't think there is the need to test for any specific annotations.

Comment (cscatolini) [reply rootfs/scheduler/resources/pod.py:126]: 👍

**Comment by felipejfc** [comment]
> @felipejfc we've decided to follow the same approach used to support tolerations (via env vars) so there would be changes only on the controller instead of changing the controller, the cli and the builder. This will allow us to fix this issue quicker, since it is already impacting our deis deployments running with the new datadog agent.

I think we should think more on the clients in this matter, as a client I really dislike the solution it's more complicated and less intuitive than doing it using a command in the cli

**Review by cscatolini — APPROVED** [review]

**Comment by cscatolini** [comment]
@felipejfc ok, I agree with you on that point but I still think we should make a quick solution available.

@Kasama we'll deploy this as it is and please open a new PR allowing annotations and tolerations to be configured using the CLI. We should make it backwards compatible so we'll use the environment variable if available, but we should override it with the info provided via CLI.

**Pull request merged**